File: implementations/php/mergesort.php
<?php function mergesort($arr){
	if(count($arr) == 1 ) return $arr;
	$mid = count($arr) / 2;
    $left = array_slice($arr, 0, $mid);
    $right = array_slice($arr, $mid);
	$left = mergesort($left);
	$right = mergesort($right);
	return merge($left, $right);
}
 
function merge($left, $right){
	$res = array();
	while (count($left) > 0 && count($right) > 0){
		if($left[0] > $right[0]){
			$res[] = $right[0];
			$right = array_slice($right , 1);
		}else{
			$res[] = $left[0];
			$left = array_slice($left, 1);
		}
	}
	while (count($left) > 0){
		$res[] = $left[0];
		$left = array_slice($left, 1);
	}
	while (count($right) > 0){
		$res[] = $right[0];
		$right = array_slice($right, 1);
	}
	return $res;
}
 
function customRandom($x)
{ $MAXI = 1000000 ; 
    return (floor($x*$x*$x*cos($x)*cos($x))) % $MAXI ;    

}

$n = ($argc > 1 ) ? $argv[1] : 100 ; 
$niters = ($argc > 2 ) ? $argv[2] : 1 ;  
for ($iter=0;$iter < $niters ;$iter++) {
	$arr = [] ; 
	for ($i=0; $i < $n ; $i++) {
	    $arr[$i]= customRandom($i);  
	
	}


	//echo implode(' ',$arr); echo "\n" ;
	$arr = mergesort($arr);
	//echo implode(' ',$arr); echo "\n" ;
}
?>

File: implementations/nim/mergesort.nim
#nimlang/nim

import 
    math ,
    strutils ,
    os 

proc merge[T](a, b: var openarray[T], left, middle, right:int) =
    let
      leftLen = middle - left
      rightLen = right - middle
    var
      l = 0
      r = leftLen
   
    for i in left .. <middle:
      b[l] = a[i]
      inc l
    for i in middle .. < right:
      b[r] = a[i]
      inc r
   
    l = 0
    r = leftLen
    var i = left
   
    while l < leftLen and r < leftLen + rightLen:
      if b[l] < b[r]:
        a[i] = b[l]
        inc l
      else:
        a[i] = b[r]
        inc r
      inc i
   
    while l < leftLen:
      a[i] = b[l]
      inc l
      inc i
    while r < leftLen + rightLen:
      a[i] = b[r]
      inc r
      inc i
   
proc mergeSort[T](a, b: var openarray[T], left, right:int) =
    if right - left <= 1: return
   
    let middle = (left + right) div 2
    mergeSort(a, b, left, middle)
    mergeSort(a, b, middle, right)
    merge(a, b, left, middle, right)
   
proc mergeSort[T](a: var openarray[T]) =
    var b = newSeq[T](a.len)
    mergeSort(a, b, 0, a.len)

proc customRandom(x:int): int =
    const MAXI = 1000000
    return  int(float(x*x*x )*  math.cos(float(x))*math.cos(float(x))) mod MAXI  



var
    n = 20
    niters = 1
    a : seq[int] 

if paramCount() > 0 : 
    n = parseInt(paramStr(1))

if paramCount() > 1 : 
    niters = parseInt(paramStr(2))

for iter in 0..(niters-1) :
    
    a = newSeq[int](n)
    for i in 0..(n-1) : 
        a[i] = customRandom(i)
    
    echo a    
    mergeSort a
    echo a